### ISSUE TITLE
Airplay stream never appears in snap clients

### ISSUE THREAD

**Issue by carsoni** [badaix/snapcast#962]
**TL;DR**
In all shairport-sync / snapserver setups I have tried the airplay device appears in IOS and in the snap server logs as "playing" and showing `onResync` but the source never appears in any snap client.
Any pointers would be gratefully received.
Let me know what other logs would be useful

Many thanks in advance
Regards
Ian Carson

**Context:**
* Multi-room audio setup
* Music server is running on Ubuntu 20.04 VM on TrueNAS12 - no audio hardware
* Music server is using MPD for music database and Cantata for testing
* Music server is running snapcast snapserver
* I have snap clients running on a latte panda, am using snapweb on windows and the snapcast client on IOS15 on iPhone13Pro
* I have and MPD client running on IOS
* MPD has an audio_ouput to the /tmp/snapfifo pipe
* Snap server uses /tmp/snapfifo as a source named "MPD"

**What's working**
* MPD source appears in all the snap clients
* Snap server outputs music to all snap clients when music is played through Cantata on the music server
* Snap server outputs music to all snap clients when MPD on the Music Server is controlled from the IOS client
* Shairport-sync named device (Airplay) appears in IOS and accepts a play command
* The play command results in a log entry of `State changed: Airplay, state: idle => playing`  

**What I've attempted** 
* I'm trying to attach Shairport-sync to the snap server as another source so that it appears in the snap clients and will allow me to stream music from my iPhone to all snap clients
* I have tried a `sudo apt install shairport-sync` and have also tried with a built shairport-sync using these switches --with-(alsa, pa, stdout, systemd, avahi, pipe, metadata, ssl=openssl)
* With both builds from above (separately of course and removing all files between attempts) I have tried with each of the following setups in turn
* Setup 1
    * shairport-sync with `output=stdout` in shairport-sync.conf, 
    * snapserver with `source=airplay:///usr/local/bin/shairport-sync?name=AirPlay&devicename=Airplay`
    * shairport-sync systemd service not running (snapserver service runs it) 
* Setup 2
    * shairport-sync with `output=pipe` and `pipe=/tmp/snapshairport` in shairport-sync.conf,
    * snapserver with `source=pipe:///tmp/snapshairport`
    * shairport-sync systemd service running
* Setup 3
    * shairport-sync with no settings set in conf file but with the service file call 
                    `/usr/local/bin/shairport-sync -o pipe -- /tmp/snapshairport` 
    * snapserver with `source=pipe:///tmp/snapshairport`

Here's a sample from my syslog (similar entries for all 3 setups)
```Dec 23 18:47:36 MusicServer snapserver[2386]: Creating TCP acceptor for address: 0.0.0.0, port: 1705
Dec 23 18:47:36 MusicServer snapserver[2386]: Creating HTTP acceptor for address: 0.0.0.0, port: 1780
Dec 23 18:47:36 MusicServer snapserver[2386]: Creating stream acceptor for address: 0.0.0.0, port: 1704
Dec 23 18:47:36 MusicServer snapserver[2386]: Number of threads: 4, hw threads: 8
Dec 23 18:47:37 MusicServer snapserver[2386]: Metadata pipe opened: /tmp/shairmeta.2386.5000
Dec 23 18:47:37 MusicServer snapserver[2386]: Waiting for metadata, retrying in 2500ms
Dec 23 18:47:37 MusicServer snapserver[2386]: ControlServer::NewConnection: 192.168.1.243
Dec 23 18:47:37 MusicServer snapserver[2386]: StreamServer::NewConnection: 192.168.1.27
Dec 23 18:47:37 MusicServer snapserver[2386]: Service 'Snapcast' successfully established.
Dec 23 18:47:37 MusicServer snapserver[2386]: onResync (MPD): 87.6973 ms
Dec 23 18:47:37 MusicServer snapserver[2386]: Hello from 50:eb:71:3f:72:30, host: latte01, v0.26.0, ClientName: Snapclient, OS: Ubuntu 20.04.3 LTS, Arch: x86_64, Protocol version: 2
Dec 23 18:47:37 MusicServer snapserver[2386]: Removing 1 inactive session(s), active sessions: 1
Dec 23 18:47:37 MusicServer snapserver[2386]: onResync (AirPlay): 70.4122 ms
Dec 23 18:47:39 MusicServer snapserver[2386]: Waiting for metadata, retrying in 2500ms
Dec 23 18:47:47 MusicServer snapserver[2386]: message repeated 3 times: [ Waiting for metadata, retrying in 2500ms]
Dec 23 18:47:49 MusicServer snapserver[2386]: Waiting for metadata, retrying in 2500ms
Dec 23 18:47:57 MusicServer snapserver[2386]: message repeated 3 times: [ Waiting for metadata, retrying in 2500ms]
Dec 23 18:48:00 MusicServer snapserver[2386]: onResync (AirPlay): 73.5716 ms
Dec 23 18:48:00 MusicServer snapserver[2386]: State changed: AirPlay, state: idle => playing
Dec 23 18:48:00 MusicServer snapserver[2386]: onStateChanged (AirPlay): playing

**Comment by efung1232** [comment]
Have you tried to do "systemctl status shairport-sync.service" to check status?
The following is from my system:
● shairport-sync.service - Shairport Sync - AirPlay Audio Receiver
   Loaded: loaded (/lib/systemd/system/shairport-sync.service; disabled; vendor preset: disabled)
   Active: active (running) since Wed 2021-12-22 13:39:15 PST; 1 day 4h ago
 Main PID: 1297 (shairport-sync)
    Tasks: 12 (limit: 4322)
   CGroup: /system.slice/shairport-sync.service
           └─1297 /usr/local/bin/shairport-sync

Warning: Journal has been rotated since unit was started. Log output is incomplete or unavailable.

**Comment by carsoni** [comment]
shairport-sync is `active (running)`

That's never been a problem and the airplay device is visible in IOS and accepts connections and shows as playing both in IOS and in the snapserver logs - but it never appears in the snap clients or snapweb

**Comment by efung1232** [comment]
This line is from my /etc/snapserver.conf: source = pipe:///tmp/shairport-sync.raw?name=Airplay
Here is my screen capture:
● snapserver.service - Snapcast server
   Loaded: loaded (/lib/systemd/system/snapserver.service; disabled; vendor preset: disabled)
   Active: active (running) since Thu 2021-12-23 17:59:53 PST; 1h 30min ago
     Docs: man:snapserver(1)
 Main PID: 2697 (snapserver)
    Tasks: 5 (limit: 4322)
   CGroup: /system.slice/snapserver.service
           └─2697 /usr/bin/snapserver --logging.sink=system --server.datadir=/var/lib/snapserver
Before I make change to use pipe, I would see status of snapserver.service showing command line execution of /usr/local/bin/shairport-sync. Now, this message is not shown.

**Comment by badaix** [comment]
Did you assign the airplay source to your client's group, e.g. in snapweb or snapdroid?

**Comment by carsoni** [comment]
Would that be a manual process - adding an entry to `/var/lib/snapserver/server.json`? Actually no AirPlay source is appearing in snapweb. That seems to be the issue - the assigned pipe is simply not appearing in the clients as a source even though the logs show that the source is "playing" and the airplay target is available in IOS and can be chosen and have music sent to it.

I was looking at attempting an entry in server.json next although it seemed strange that the `/tmp/snapfifo` pipe assigned to MPD in `snapserver.conf` gets picked up automatically.

 I was working under the assumption that the `/tmp/snapshairport` pipe I assigned to shairport in `snapserver.conf` would also be written in automatically. 

However I'm not seeing it in there. Is it valid to add a manual entry?

**Comment by efung1232** [comment]
In your first message, step 2, snapserver with source=pipe:///tmp/snapshairport
I am not sure if "?name=you-name-it" is required to be appended to "source=pipe:///tmp/snapshairport".

**Comment by badaix** [comment]
yes, the name is mandatory, otherwise the server will immediately stop with an error message.
So you don't see the source when clicking on the top left label `default: playing`? Just in case, some users are confused with the Snapweb ui.

**Comment by carsoni** [comment]
Aha!  What a beautiful little Easter Egg!  :-)

AirPlay now visible. Thanks badaix

Currently sound from the AirPlay connection is a little scratchy but I suspect that AirPlay needs to be 44100:16:2 not the 48000:16:2 I had set as a general default in snapserver.conf. A problem for tomorrow (well probably Sunday!)

Have a great Christmas and thanks for the help - it's much appreciated
Regards
Ian Carson

**Comment by badaix** [comment]
Well, some call it an easter egg, others a bad UI design, I call it my first web app prototype ;)
